packages/expo-camera/ios/CameraViewNextModule.swift

```diff
@@ -0,0 +1,247 @@
+        throw Exceptions.SimulatorNotSupported()
+        #else
+        view.stopRecording()
+        #endif
+      }.runOnQueue(.main)
+    }
+
+    AsyncFunction("launchModernScanner") { (options: VisionScannerOptions?, promise: Promise) in
+      if #available(iOS 16.0, *) {
+        Task {
+          try await MainActor.run {
+            let delegate = VisionScannerDelegate(handler: self)
+            scannerContext = ScannerContext(promise: promise, delegate: delegate)
+            launchModernScanner(with: options)
+          }
+        }
+      }
+    }
+
+    AsyncFunction("getCameraPermissionsAsync") { (promise: Promise) in
+      EXPermissionsMethodsDelegate.getPermissionWithPermissionsManager(
+        self.appContext?.permissions,
+        withRequester: CameraOnlyPermissionRequester.self,
+        resolve: promise.resolver,
+        reject: promise.legacyRejecter
+      )
+    }
+
+    AsyncFunction("requestCameraPermissionsAsync") { (promise: Promise) in
+      EXPermissionsMethodsDelegate.askForPermission(
+        withPermissionsManager: self.appContext?.permissions,
+        withRequester: CameraOnlyPermissionRequester.self,
+        resolve: promise.resolver,
+        reject: promise.legacyRejecter
+      )
+    }
+
+    AsyncFunction("getMicrophonePermissionsAsync") { (promise: Promise) in
+      EXPermissionsMethodsDelegate.getPermissionWithPermissionsManager(
+        self.appContext?.permissions,
+        withRequester: CameraMicrophonePermissionRequester.self,
+        resolve: promise.resolver,
+        reject: promise.legacyRejecter
+      )
+    }
+
+    AsyncFunction("requestMicrophonePermissionsAsync") { (promise: Promise) in
+      EXPermissionsMethodsDelegate.askForPermission(
+        withPermissionsManager: self.appContext?.permissions,
+        withRequester: CameraMicrophonePermissionRequester.self,
+        resolve: promise.resolver,
+        reject: promise.legacyRejecter
+      )
+    }
+  }
+
+  @available(iOS 16.0, *)
+  @MainActor private func launchModernScanner(with options: VisionScannerOptions?) {
+    let symbologies = options?.toSymbology()
+    let controller = DataScannerViewController(
+      recognizedDataTypes: [.barcode(symbologies: symbologies ?? [.qr, .dataMatrix])],
+      isPinchToZoomEnabled: options?.isPinchToZoomEnabled ?? false,
+      isGuidanceEnabled: options?.isGuidanceEnabled ?? false,
+      isHighlightingEnabled: options?.isHighlightingEnabled ?? false
+    )
+
+    if let delegate = scannerContext?.delegate as? VisionScannerDelegate {
+      controller.delegate = delegate
+    }
+
+    appContext?.utilities?.currentViewController()?.present(controller, animated: true) {
+      try? controller.startScanning()
+    }
+  }
+
+  func onItemScanned(result: [String: Any]) {
+    sendEvent("onModernBarcodeScanned", result)
+  }
+}
+
+private func takePictureForSimulator(
+  _ appContext: AppContext?,
+  _ view: CameraView,
+  _ options: TakePictureOptions,
+  _ promise: Promise
```

Comment:
Do we actually need a promise in this function? 😅

Reply:
Not every code path returns. Do we need to do anything if it's the simulator?
```swift
  if options.fastMode {
    view.onPictureSaved([
      "data": result,
      "id": options.id
    ])
  } else {
    promise.resolve(result)
  }
```